XenAdmin/Wizards/PatchingWizard/PatchingWizard_AutomatedUpdatesPage.cs

```diff
@@ -431,30 +418,21 @@ private void WorkerDoWork(object sender, DoWorkEventArgs doWorkEventArgs)
                 //this pool failed, we will stop here, but try to remove update files at least
                 try
                 {
-                    var positionOfFailedAction = bgw.PlanActions.IndexOf(action);
-
                     // can try to clean up the host after a failed PlanAction from bgw.PlanActions only
-                    if (positionOfFailedAction != -1 && !(action is DownloadPatchPlanAction || action is UploadPatchToMasterPlanAction))
+                    if (!(action is DownloadPatchPlanAction || action is UploadPatchToMasterPlanAction || action is RemoveUpdateFileFromMasterPlanAction))
                     {
-                        int pos = positionOfFailedAction;
-
-                        if (!(bgw.PlanActions[pos] is RemoveUpdateFileFromMasterPlanAction)) //can't do anything if the remove action has failed
+                        int pos = 0;
+                        while (++pos < bgw.FinalActions.Count)
                         {
-                            while (++pos < bgw.PlanActions.Count)
-                            {
-                                if (bgw.PlanActions[pos] is RemoveUpdateFileFromMasterPlanAction) //find the next remove
-                                {
-                                    bgw.PlanActions[pos].Run();
-                                    break;
-                                }
-                            }
+                            var finalAction = bgw.FinalActions[pos];
+                            if (finalAction is RemoveUpdateFileFromMasterPlanAction)

```

Comment:
Indeed; the error is handled, but I agree it should be made a bit smarter to avoid needless failures.

Reply:
Actually, I was wrong: the action is called, but it checks the mappings of uploaded patches and returns if it can't find the requested patch there.
A smaller issue though is that the cleanup is not run as a plan action so the user is not aware of it happening.